lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2689,12 +2722,97 @@ void ModuleEmitter::emitHWGeneratedModule(HWModuleGeneratedOp module) {
+// regs, or ports, with legalized names, so we can lookup up the names through
+// the IR.
+void ModuleEmitter::emitBind(BindOp op) {
+  InstanceOp inst = op.getReferencedInstance();
+
+  HWModuleOp parentMod = inst->getParentOfType<hw::HWModuleOp>();
+  auto parentVerilogName = getVerilogModuleNameAttr(parentMod);
+  verifyModuleName(op, parentVerilogName);
+
+  Operation *childMod = inst.getReferencedModule();
+  auto childVerilogName = getVerilogModuleNameAttr(childMod);
+  verifyModuleName(op, childVerilogName);
+
+  indent() << "bind " << parentVerilogName.getValue() << " "
+           << childVerilogName.getValue() << ' ' << inst.getName() << " (";
+
+  SmallVector<ModulePortInfo> parentPortInfo = parentMod.getPorts();
+  SmallVector<ModulePortInfo> childPortInfo = getModulePortInfo(childMod);
+
+  // Get the max port name length so we can align the '('.
+  size_t maxNameLength = 0;
+  for (auto &elt : childPortInfo) {
+    maxNameLength = std::max(maxNameLength, elt.getName().size());
+  }
+
+  // Emit the argument and result ports.
+  auto opArgs = inst.inputs();
+  auto opResults = inst.getResults();
+  bool isFirst = true; // True until we print a port.
+  for (auto &elt : childPortInfo) {
+    // Figure out which value we are emitting.
+    Value portVal = elt.isOutput() ? opResults[elt.argNum] : opArgs[elt.argNum];
+    bool isZeroWidth = isZeroBitType(elt.type);
+
+    // Decide if we should print a comma.  We can't do this if we're the first
+    // port or if all the subsequent ports are zero width.
+    if (!isFirst) {
+      bool shouldPrintComma = true;
+      if (isZeroWidth) {
+        shouldPrintComma = false;
+        for (size_t i = (&elt - childPortInfo.data()) + 1,
+                    e = childPortInfo.size();
+             i != e; ++i)
+          if (!isZeroBitType(childPortInfo[i].type)) {
+            shouldPrintComma = true;
+            break;
+          }
+      }
+
+      if (shouldPrintComma)
+        os << ',';
+    }
+    os << "\n";
+
+    // Emit the port's name.
+    indent();
+    if (!isZeroWidth) {
+      // If this is a real port we're printing, then it isn't the first one. Any
+      // subsequent ones will need a comma.
+      isFirst = false;
+      os << "  ";
+    } else {
+      // We comment out zero width ports, so their presence and initializer
+      // expressions are still emitted textually.
+      os << "//";
+    }
+
+    os << '.' << elt.getName();
+    os.indent(maxNameLength - elt.getName().size()) << " (";
+
+    // Emit the value as an expression.
+    auto name = getNameRemotely(portVal, parentPortInfo);
+    if (name.empty()) {
+      // Non stable names will come from expressions.  Since we are lowering the
+      // instance also, we can ensure that expressions feeding bound instances
+      // will be lowered consistently to verilog-namable entities.
+      os << childVerilogName.getValue() << '_' << inst.getName() << '_'
+         << elt.getName() << ')';
```

Comment:
It looks like this logic will not work if the tapped thing is a temporary.  E.g., `%0 = sv.read_inout %x` will result in some weirdness in the bind like `Bar_Bar_a`. I'll open an issue.